server/src/main/java/org/opensearch/search/aggregations/bucket/terms/GlobalOrdinalsStringTermsAggregator.java

```diff
@@ -136,16 +144,92 @@ public GlobalOrdinalsStringTermsAggregator(
+     @param ordCountConsumer A consumer to accept collected term frequencies
+     @return A no-operation LeafBucketCollector implementation, since collection is complete
+     @throws IOException If an I/O error occurs during reading
+     */
+    LeafBucketCollector termDocFreqCollector(
+        LeafReaderContext ctx,
+        SortedSetDocValues globalOrds,
+        BiConsumer<Long, Integer> ordCountConsumer
+    ) throws IOException {
+        if (weight == null || weight.count(ctx) != ctx.reader().maxDoc()) {
+            // Weight not assigned or top-level query does not match all docs in the segment.
+            return null;
+        }
+
+        if (weight.count(ctx) == 0) {
+            // No documents matches top level query on this segment, we can skip the segment
+            return LeafBucketCollector.NO_OP_COLLECTOR;
+        }
+
+        Terms segmentTerms = ctx.reader().terms(this.fieldName);
+        if (segmentTerms == null) {
+            // Field is not indexed.
+            return null;
+        }
+
+        NumericDocValues docCountValues = DocValues.getNumeric(ctx.reader(), DocCountFieldMapper.NAME);
+        if (docCountValues.nextDoc() != NO_MORE_DOCS) {
+            // This segment has at least one document with the _doc_count field.
+            return null;
+        }
+
+        TermsEnum indexTermsEnum = segmentTerms.iterator();
+        BytesRef indexTerm = indexTermsEnum.next();
+        TermsEnum globalOrdinalTermsEnum = globalOrds.termsEnum();
+        BytesRef ordinalTerm = globalOrdinalTermsEnum.next();
+
+        while (indexTerm != null && ordinalTerm != null) {
+            int compare = indexTerm.compareTo(ordinalTerm);
+            if (compare == 0) {
+                if (acceptedGlobalOrdinals.test(globalOrdinalTermsEnum.ord())) {
+                    ordCountConsumer.accept(globalOrdinalTermsEnum.ord(), indexTermsEnum.docFreq());
+                }
+                indexTerm = indexTermsEnum.next();
+                ordinalTerm = globalOrdinalTermsEnum.next();
+            } else if (compare < 0) {
+                indexTerm = indexTermsEnum.next();
+            } else {
+                ordinalTerm = globalOrdinalTermsEnum.next();
+            }
+        }
+        return LeafBucketCollector.NO_OP_COLLECTOR;
+    }
+
     @Override
     public LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
         SortedSetDocValues globalOrds = valuesSource.globalOrdinalsValues(ctx);
         collectionStrategy.globalOrdsReady(globalOrds);
+
+        if (collectionStrategy instanceof DenseGlobalOrds && sub == LeafBucketCollector.NO_OP_COLLECTOR) {
+            LeafBucketCollector termDocFreqCollector = termDocFreqCollector(
+                ctx,
+                globalOrds,
+                (ord, docCount) -> incrementBucketDocCount(collectionStrategy.globalOrdToBucketOrd(0, ord), docCount)
+            );
+            if (termDocFreqCollector != null) {
```

Comment:
Better to add a debug log here say the optimization kicks in

Reply:
I don't see much debug logs in the search related code-bases. I'm skeptic if adding debug can have minor performance implications.

Reply:
In practice, log4j does a pretty good job of returning quickly when it's not logging. @bowenlan-amzn's suggestion is reasonable, IMO. (I would remove the logging I think after some point in the future, as this becomes the "normal" behavior.)

Also, the number of segments per shard should be in the double digits, never in the thousands. (We should raise floorSegmentMB and dynamically adjust max segment size to produce a constant number of equally-sized segments, IMO, but that's a separate thing.)

Reply:
Sure.

In practice, debug logging (even at thousands of shards) is unlikely to impact performance (since you're probably not enabling debug logging for your whole cluster), but it's really not that important.